File: 2023/18_LavaductLagoon/lagoon.py

# ======================================================================
# Lavaduct Lagoon
#   Advent of Code 2023 Day 18 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                           l a g o o n . p y
# ======================================================================
"Lagoon for the Advent of Code 2023 Day 18 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

from plan import Plan
# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
DELTA = {
    'U': (-1, 0),
    'D': (1, 0),
    'R': (0, 1),
    'L': (0, -1),
}

MAP_TRENCH = '#'
MAP_GROUND = '.'
MAP_DOUBLE = ".#."

# ======================================================================
#                                                                 Lagoon
# ======================================================================


class Lagoon(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Lavaduct Lagoon"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.plan = None
        self.trench = {}
        self.corners = []
        self.distance = 0

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self.plan = Plan(text=text, part2=part2)

    def execute_plan(self):
        "Execute the plan"

        # 1. Start at the very beginning
        loc = (0, 0)
        self.trench[loc] = "000000"
        self.corners.append(loc)
        distance = 0

        # 2. Follow each step of the plan
        for step in self.plan:

            # 3. Get the step specifics
            action, meters, hex_color = step.specs()
            delta = DELTA[action]

            # 4. Dig in the indicated direction
            if not self.part2:
                for _ in range(meters):
                    loc = (loc[0] + delta[0], loc[1] + delta[1])
                    self.trench[loc] = hex_color
            else:
                loc = (loc[0] + meters * delta[0], loc[1] + meters * delta[1])
                distance += meters

            # 5. Keep track of the corners
            self.corners.append(loc)

        # 6. Remember the lenght of the trench
        if not self.part2:
            self.distance = len(self.trench)
        else:
            self.distance = distance

        # 7. Return the trench distance (mainly for testing)
        return self.distance

    def get_dimensions(self):
        "Determine the outside dimensions of the lagoon"

        # 1. Start with nothing
        min_row = 99999999
        min_col = 99999999
        max_row = 0
        max_col = 0

        # 2. Loop for all of the trench locations
        for trench in self.trench:

            # 3. Adjust the bounding box
            row, col = trench
            min_row = min(min_row, row)
            max_row = max(max_row, row)
            min_col = min(min_col, col)
            max_col = max(max_col, col)

        # 4. Return the bounding box
        return min_row, min_col, max_row, max_col

    def __str__(self):
        "Returns a map of the trench"

        # 1. Start with nothing
        rows = []

        # 2. Get the dimensions of the trench
        min_row, min_col, max_row, max_col = self.get_dimensions()

        # 3. Loop for all the rows
        for rindx in range(min_row, max_row + 1):

            # 4. Loop for all the columns
            this_row = self.map_this_row(rindx, min_col, max_col)

            # 5. Add the the newly mapped row
            rows.append(this_row)

        # 6. return the trench map
        return '\n'.join(rows)

    def map_this_row(self, rindx, min_col, max_col):
        "Map this row"

        # 1. Start with nothing
        this_row = []

        # 2. Map this row
        for cindx in range(min_col, max_col + 1):
            if (rindx, cindx) in self.trench:
                this_row.append(MAP_TRENCH)
            else:
                this_row.append(MAP_GROUND)

        # 4. Return the row as a string
        return ''.join(this_row)

    def row_interior(self, rindx, min_col, max_col):
        "Return the number of internal spaces in this lagoon row"

        # 1. Start with nothing
        result = 0
        parity = False

        # 2. Map this row
        this_row = MAP_GROUND + self.map_this_row(rindx, min_col, max_col) + MAP_GROUND

        # 3. Replace pairs of trenches with a single trench
        for pounds in range(2, 1 + max_col - min_col, 2):
            repl = MAP_GROUND + (MAP_TRENCH * pounds) + MAP_GROUND
            this_row = this_row.replace(repl, MAP_DOUBLE)
        this_row = this_row.strip(MAP_GROUND)
        # print(f"{rindx}: {this_row}")

        # 4. Loop for all the columns in the adjusted row
        for char in this_row:

            # 5. If this column is part of the trench, adjust parity
            if char == MAP_TRENCH:
                parity = not parity

            # 6. Else the column is part of the lagoon if parity is True
            elif parity:
                result += 1

        # 7. Return the number of interier spaces in this row
        return result

    def shoelace_area(self):
        "Return the area computed by the Shoelace Formula (since parity isn't working)"

        # 1. Start with nothing
        result = 0

        # 2. Loop for every corner
        for cindx, corner in enumerate(self.corners):

            # 3. Calculate the signed area contribution of the current segment
            area = (corner[0]
                    * (self.corners[cindx - 1][1] -
                     self.corners[(cindx + 1) % len(self.corners)][1]))
            result += area

        # 4. Normalize the area
        return abs(result) // 2

    def cubic_meters(self):
        "Return the number of cubic meters in the dug out lagoon"

        # 1. Get the area of the lagoon
        area = self.shoelace_area()

        # 2. Use Pick's theorem to calculate the number of interior points
        # https://en.wikipedia.org/wiki/Pick%27s_theorem
        interior = area - self.distance // 2 + 1

        # 4. Normalize the area
        return interior + self.distance

    def old_cubic_meters(self):
        "Return the number of cubic meters in the dug out lagoon"

        # 1. Start with nothing
        result = 0

        # 2. Get the dimensions of the trench
        min_row, min_col, max_row, max_col = self.get_dimensions()

        # 3. Loop for all the rows
        for rindx in range(min_row, max_row + 1):

            # 4. Get the number of interior spaces in this row
            row_spaces = self.row_interior(rindx, min_col, max_col)

            # 5. Add these to the total
            result += row_spaces

        # 6. Return the total size of lagoon (interior and trench)
        return result + len(self.trench)

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                        l a g o o n . p y                       end
# ======================================================================

File: 2023/18_LavaductLagoon/lagoon.go
// ======================================================================
// Lavaduct Lagoon
//   Advent of Code 2023 Day 18 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         l a g o o n . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 18 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const (
	LOC   = 100000000
	LOCP  = LOC * 10000000
	UP    = 'U'
	DOWN  = 'D'
	RIGHT = 'R'
	LEFT  = 'L'
	DIRS  = "RDLU"
)

var (
	DELTA = map[rune][]int{
		UP:    {-1, 0},
		DOWN:  {1, 0},
		RIGHT: {0, 1},
		LEFT:  {0, -1},
	}
)

// ----------------------------------------------------------------------
//
//	Utility
//
// ----------------------------------------------------------------------
func nextLoc(row int, col int, direction rune, length int) (int, int) {
	nextRow := row + DELTA[direction][0]*length
	nextCol := col + DELTA[direction][1]*length
	return nextRow, nextCol
}

// ======================================================================
//                                                                 Lagoon
// ======================================================================

// Lagoon ... Object for Lavaduct Lagoon
type Lagoon struct {
	Part2   bool
	Text    []string
	Corners [][]int
	Trench  int
}

// NewLagoon ... Create a Lagoon object
func NewLagoon(part2 bool, text []string) (*Lagoon, error) {

	// 1. Set the initial values
	p := new(Lagoon)
	p.Part2 = part2
	p.Text = text

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Lagoon object
	return p, nil
}

// processText ... Assign values from text
func (p *Lagoon) processText(text []string) error {

	// 1. Start in the upper left
	row := 0
	col := 0
	p.Corners = make([][]int, 1, 2+len(p.Text))
	p.Corners[0] = []int{row, col}

	// 1. Loop for each line of the text
	for indx, line := range text {

		// 2. Decode the line
		parts := strings.Fields(line)
		if len(parts) != 3 {
			return fmt.Errorf("bad format in line %d: '%s'", indx, line)
		}
		direction := parts[0]
		if len(direction) != 1 || !strings.ContainsAny(direction, DIRS) {
			return fmt.Errorf("invalid direction '%s' in line %d: '%s'", direction, indx, line)
		}
		strMeters := parts[1]
		meters, err := strconv.ParseInt(strMeters, 10, 64)
		if err != nil {
			return fmt.Errorf("invalid meters '%s' in line %d: '%s'", strMeters, indx, line)
		}
		color := parts[2]
		if len(color) != 9 {
			return fmt.Errorf("invalid color length '%s' in line %d: '%s'", color, indx, line)
		}
		if !strings.HasPrefix(color, "(#") {
			return fmt.Errorf("invalid color prefix '%s' in line %d: '%s'", color, indx, line)
		}
		if !strings.HasSuffix(color, ")") {
			return fmt.Errorf("invalid color suffix '%s' in line %d: '%s'", color, indx, line)
		}
		color = color[2:8]

		// 3. Part 2 needs to decode the color
		if p.Part2 {
			meters, err = strconv.ParseInt(color[0:5], 16, 64)
			if err != nil {
				return fmt.Errorf("invalid color meters '%s' in line %d: '%s'", color, indx, line)
			}
			dirn, err := strconv.ParseInt(string(color[5]), 16, 64)
			if err != nil {
				return fmt.Errorf("invalid color number '%s' in line %d: '%s'", string(color[5]), indx, line)
			}
			if dirn < 0 || dirn > 3 {
				return fmt.Errorf("invalid color value '%s' in line %d: '%s'", string(color[5]), indx, line)
			}
			direction = string(DIRS[dirn])
		}

		// 4. Calculate next corner and save it
		row, col = nextLoc(row, col, rune(direction[0]), int(meters))
		p.Corners = append(p.Corners, []int{row, col})
		p.Trench = p.Trench + int(meters)
	}

	// 4. And then back to the start
	if row != 0 || col != 0 {
		return fmt.Errorf("dig plan did not return to origin")
	}

	// 5. Return success
	return nil
}

// ShoelaceArea ... Return the area of the lagoon
//
//	from https://en.wikipedia.org/wiki/Shoelace_formula
func (p *Lagoon) ShoelaceArea() int {

	// 1. Start with nothing
	corners := len(p.Corners)
	result := 0

	// 2. Loop for the remaining corners
	for indx := 0; indx < corners; indx++ {
		jndx := (indx + 1) % corners

		// 3. Get the row and column of this corner
		rcIndx := p.Corners[indx]
		rcJndx := p.Corners[jndx]

		// 4. Add in this corner
		result = result + (rcIndx[0] * rcJndx[1]) - (rcJndx[0] * rcIndx[1])
	}

	// 6. Return the area
	return int(math.Abs(float64(result / 2)))
}

// LagoonArea ... Total size of the lagoon
func (p *Lagoon) LagoonArea() int {

	// 1. Get the area by the shoestring method
	slArea := p.ShoelaceArea()

	// 2. Get the number of interior points
	interior := 1 + slArea - p.Trench/2

	// 3. Return the total size of the lagoon
	print("sl=", slArea, " int=", interior, " trench=", p.Trench, " total=", interior+p.Trench, "\n")
	return interior + p.Trench
}

// PartOne ... Returns the solution for part one
func (p *Lagoon) PartOne(verbose bool, limit int) string {

	// 1. Return the solution for part one
	return fmt.Sprint(p.LagoonArea())
}

// PartTwo ... Returns the solution for part two
func (p *Lagoon) PartTwo(verbose bool, limit int) string {

	// 1. Return the solution for part two
	return fmt.Sprint(p.LagoonArea())
}

// ======================================================================
// end                        l a g o o n . g o                       end
// ======================================================================
